### PR TITLE
PHOENIX-5117 : Return the count of rows scanned in HBase

### PR DESCRIPTION
**Pull request by palashc**

### CODE REVIEW

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java [hunk 1]
```diff
@@ -125,21 +125,25 @@ public class QueryCompiler {
...
-        this(statement, select, resolver, new BindManager(statement.getParameters()), targetColumns, parallelIteratorFactory, sequenceManager, projectTuples, optimizeSubquery, dataPlans);
+        this(statement, select, resolver, new BindManager(statement.getParameters()), targetColumns, parallelIteratorFactory, sequenceManager, projectTuples, optimizeSubquery, dataPlans, null);
     }
 
-    public QueryCompiler(PhoenixStatement statement, SelectStatement select, ColumnResolver resolver, BindManager bindManager, List<? extends PDatum> targetColumns, ParallelIteratorFactory parallelIteratorFactory, SequenceManager sequenceManager, boolean projectTuples, boolean optimizeSubquery, Map<TableRef, QueryPlan> dataPlans) throws SQLException {
-        this.statement = statement;
+    public QueryCompiler(PhoenixStatement statement, SelectStatement select, ColumnResolver resolver, List<? extends PDatum> targetColumns, ParallelIteratorFactory parallelIteratorFactory, SequenceManager sequenceManager, boolean projectTuples, boolean optimizeSubquery, Map<TableRef, QueryPlan> dataPlans, StatementContext parentContext) throws SQLException {
```
Comment (tkhurana) [inline phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:142]: Don't change public API signatures. It causes API compatibility issues.

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java [hunk 2]
```diff
@@ -705,6 +717,9 @@ protected QueryPlan compileSubquery(
                     statement,
                     queryPlan);
         }
+        if (parentContext != null) {
+            parentContext.addSubStatementContext(queryPlan.getContext());
```
Comment (tkhurana) [inline phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:721]: What is the reasoning behind this change?

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        }
+        return getRowsScanned(rs);
+    }
+
+    private long getRowsScanned(ResultSet rs) {
+        if (!(rs instanceof PhoenixResultSet)) {
+            return -1;
+        }
+        PhoenixResultSet phoenixResultSet = (PhoenixResultSet) rs;
+        Map<String, Map<MetricType, Long>> metrics = phoenixResultSet.getReadMetrics();
```
Comment (tkhurana) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:282]: Better to use `PhoenixRuntime.getRequestReadMetricInfo`

Comment (tkhurana) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:282]: Also, check the tableName which is the key in the map to ensure we are only extracting the relevant metrics.

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        Map<String, String> props = Maps.newHashMapWithExpectedSize(1);
+        setUpTestDriver(new ReadOnlyProps(props));
+    }
+
+    @Test
+    public void testSinglePrimaryKey() throws Exception {
+        Properties props = new Properties();
+        props.put(QueryServices.COLLECT_REQUEST_LEVEL_METRICS, "true");
+        // force many rpc calls
+        props.put(QueryServices.SCAN_CACHE_SIZE_ATTRIB, "10");
```
Comment (tkhurana) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:56]: I see this code being repeated in multiple tests. We could move it to the setup method.

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        Connection conn = DriverManager.getConnection(getUrl(), props);
+        String tableName = generateUniqueName();
+        PhoenixStatement stmt = conn.createStatement().unwrap(PhoenixStatement.class);
+        stmt.execute("CREATE TABLE " + tableName
```
Comment (tkhurana) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:60]: Can you also add a testcase with skip scan. Like `WHERE A IN (20, 45, 68, 3)`

**Comment by tkhurana** [comment]
@palashc Another test I want you to write is where there are delete markers in the beginning of the table and then you run a query like this `SELECT * FROM <TABLE> LIMIT 1`. It should skip the delete markers till it finds the first valid row.  You could run this query in a loop where in every iteration you could delete more rows which have to be skipped in the subsequent SELECT query.

File: phoenix-core-server/src/main/java/org/apache/phoenix/cache/aggcache/SpillableGroupByCache.java [hunk 6]
```diff
@@ -365,6 +366,10 @@ public void close() throws IOException {
                 }
             }
 
+            public boolean next(List<Cell> result, ScannerContext scannerContext) throws IOException {
+                return next(result);
```
Comment (tkhurana) [inline phoenix-core-server/src/main/java/org/apache/phoenix/cache/aggcache/SpillableGroupByCache.java:370]: Why is the scannerContext being silently ignored here ?

File: phoenix-core-client/pom.xml [hunk 7]
```diff
@@ -438,5 +438,9 @@
       <groupId>org.hdrhistogram</groupId>
       <artifactId>HdrHistogram</artifactId>
     </dependency>
+    <dependency>
```
Comment (tkhurana) [inline phoenix-core-client/pom.xml:441]: Is it OK to add this dependency ? @stoty can you confirm ?

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java (hunk 4 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java]: done

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java (hunk 5 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:59]: @tkhurana done

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+        assertEquals(79, count5);
+
+        // pk1 and pk2, group by, ordered
+        long count6 = countRowsScannedFromSql(stmt, "SELECT B, SUM(A), SUM(Z) FROM " + tableName
+                + " WHERE A >= 2 AND B >= 3 GROUP BY B ORDER BY B DESC");
+        assertEquals(79, count6);
+    }
+
+    @Test
+    public void testQueryWithDeleteMarkers() throws Exception {
```
Comment (palashc) [inline phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:177]: @tkhurana Added this test for deletes, let me know if this makes sense.

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java (hunk 3 above)
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java]: done

File: phoenix-core-client/pom.xml (hunk 7 above)
Comment (virajjasani) [reply phoenix-core-client/pom.xml:441]: Having phoenix-core-client depend on hbase-server is really bad practice because it ultimately kills the whole purpose of having two separate phoenix-core modules, one for client and one of server.

But the bigger problem here is that we use the same iterator class at both client and server side, and depending on which side uses it, the context like this can be used.

It would have been much simpler if we had iterators of type `ResultIterator` only being used by clients and scanners of type `BaseRegionScanner` only being used by server. Then we could make `ScannerContext` available to only scanners of type `BaseRegionScanner`.

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java [hunk 9]
```diff
@@ -465,6 +475,10 @@ public boolean next(List<Cell> results) throws IOException {
                         }
                     }
                     tuple = nextTuple;
+                    if (regionScannerContext != null) {
+                        ScannerContextUtil.updateMetrics(regionScannerContext, scannerContext);
+                        regionScannerContext = null;
+                    }
```
Comment (virajjasani) [inline phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:481]: I wonder if it is still possible to not pass the ScannerContext to iterators above and still use it here?
Probably we can, since we are not really updating any context in `OrderedResultIterator` and `OffsetResultIterator`? Or i am missing something?

File: phoenix-core-client/pom.xml (hunk 7 above)
Comment (virajjasani) [reply phoenix-core-client/pom.xml:441]: Now i think probably we might still be able to get around this by not passing scanner context to `OrderedResultIterator` and `OffsetResultIterator`?

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java (hunk 9 above)
Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:481]: @virajjasani how would we pass the context from RegionScannerResultIterator to NonAggregateRegionScannerFactory?

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java (hunk 2 above)
Comment (palashc) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:721]: @tkhurana This is to keep track of the Context for nested/inner Select queries. During `getReadMetrics` in `PhoenixResultSet`, we combine the metrics from the inner contexts with the outer context.

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java (hunk 1 above)
Comment (palashc) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:142]: @tkhurana The existing overloaded constructors did not change, they just provide a null value for the new field `parentContext`.

Comment (tkhurana) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:142]: @palashc But one constructor API which is public is  changing right ?

Comment (palashc) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:142]: @tkhurana Only the one which sets all the fields and only has usage in the class itself.

Comment (tkhurana) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:142]: We have seen in the past where other applications are using the public API. Simply searching in the Phoenix code space is not enough. The rule of thumb is don't change the signature of public API.

File: phoenix-core-server/src/main/java/org/apache/phoenix/cache/aggcache/SpillableGroupByCache.java (hunk 6 above)
Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/cache/aggcache/SpillableGroupByCache.java:370]: @tkhurana I am not sure what the implementation should be for scanners returned by SpillableGroupByCache and InMemoryGroupByCache

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java (hunk 1 above)
Comment (palashc) [reply phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:146]: Got it, understood. I have made the fix to not change any of the 5 existing constructors.

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java [hunk 10]
```diff
@@ -298,6 +303,24 @@ static OrderedResultIterator deserializeFromScan(Scan scan, RegionScanner s,
         }
     }
 
+    private static class IteratorAndContext {
```
Comment (virajjasani) [inline phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:306]: @palashc you think this is what seems tricky? Let me take a look next week in detail.

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:306]: @virajjasani Yes, exactly. Maybe there is another way to refactor here.

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java (hunk 9 above)
Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:500]: Tried to address in latest commit

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseRegionScanner.java [hunk 11]
```diff
@@ -42,9 +42,7 @@ public boolean isFilterDone() {
     public abstract boolean next(List<Cell> results) throws IOException;
 
     @Override
-    public boolean next(List<Cell> result, ScannerContext scannerContext) throws IOException {
```
Comment (stoty) [inline phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseRegionScanner.java:45]: Why is that ?
I can see this implemented in child classes.

File: phoenix-core-client/pom.xml (hunk 7 above)
Comment (stoty) [reply phoenix-core-client/pom.xml]: Can you point me where this is a problem in the code ?
There are too many change files for me to check easily.

Comment (palashc) [reply phoenix-core-client/pom.xml]: @stoty initial implementation had introduced a dependency on hbase-server in phoenix-client since I was passing ScannerContext through `OrderedResultIterator` and `OffsetResultIterator`, which was not acceptable. I have now tried an alternate implementation where I avoid doing this.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseRegionScanner.java (hunk 11 above)
Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseRegionScanner.java:45]: Declared next with scannerContext as abstract to make sure all child classes implement it.

File: phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java [hunk 12]
```diff
@@ -36,4 +36,28 @@ public static void incrementSizeProgress(ScannerContext sc, List<Cell> cells) {
...
+        sc.updateTimeProgress();
+    }
+
+    /**
+     * Increment the heapSize progress to a large value for server paging.
+     * Not using Long.MAX_VALUE since that can cause overflow.
+     */
+    public static void incrementHeapProgressForPaging(ScannerContext sc) {
+        sc.incrementSizeProgress(0, Long.MAX_VALUE/2);
+    }
```
Comment (virajjasani) [inline phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: @kadirozde @palashc This is nice trick. Is `Long.MAX_VALUE/2` guaranteed to cause large enough size for HBase to return response?

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: Yes, 4096 PB will be beyond any practical heap size for a very long time.

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: However, you have a good point. @palashc, the default heap size is -1, which means it is not enforced. What happens in that case?

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: @palashc, we may have to set the heap limit to Long.MAX_VALUE/2 on the scanner context on the way down and increment heap progress by Long.MAX_VALUE/2 when dummy result is received. Also we need to make sure that this manipulation of the scanner context is done only once for a given scanner.

Comment (virajjasani) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: @palashc it seems we have better alternative i.e. instead of size limit, we can enforce time limit reached by using

```
  // Set this to true will have the same behavior with reaching the time limit.
  // This is used when you want to make the current RSRpcService.scan returns immediately. For
  // example, when we want to switch from pread to stream, we can only do it after the rpc call is
  // returned.
  private boolean returnImmediately;
```

```
  void returnImmediately() {
    this.returnImmediately = true;
  }
```

This enforces `timeLimitReached`. HBase uses three factors to determine whether to return immediately:
1. sizeLimitReached
2. timeLimitReached
3. resultsLimitReached

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java:62]: @palashc, the above by @virajjasani is a much simpler and better approach. Please verify that it works.

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java]: Made the change in the latest commit after verifying using the new IT I added in ServerPagingIT. 
Thanks @virajjasani , great find!

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/hadoop/hbase/regionserver/ScannerContextUtil.java]: @kadirozde @virajjasani Please let me know if you would be able to review the rest of the changes in this PR regarding scannerContext.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java [hunk 13]
```diff
@@ -97,4 +152,10 @@ public RegionScanner getNewRegionScanner(Scan scan) throws IOException {
             throw new DoNotRetryIOException(e);
         }
     }
+
+    public static class ScannerContextNextNotImplementedException extends IOException {
```
Comment (kadirozde) [inline phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: Why is this necessary?

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: It can be used to inform DelegateRegionScanner that a lower level does not implement next/nextRaw with scannerContext, for example in this change in HashJoinRegionScanner: https://github.com/apache/phoenix/pull/1936/files#diff-77ea25574d36a1657d82cedf3384ebca3d8b11ac7a0de74dd5bcb18c59069afcR352

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: Why does HashJoinRegionScanner not support it? Should all layers support this now without exception?

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: `HashJoinRegionScanner.nextRaw` supports it. Should `next()` also support it?

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: Yes, all should support it from now on

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java:156]: What should be the implementation for next in HashJoinRegionScanner, with and without ScannerContext? Should it just call the corresponding nextRaw?

Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/DelegateRegionScanner.java]: I made the change in the latest commit.

**Comment by palashc** [comment]
Latest test report: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1936/29/testReport/

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java (hunk 10 above)
Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java:321]: I did not understand what we are achieving here. OrderedResultIterator is used in both client and server side. So, we need to add the next() with ScannerContext method to OrderedResultIterator too. This method will be used on the server side. The scanner context should not serialized/deserialized. We also need to pass through ScannerContext context for the other result iterators used on the server side (for example, OffsetResultIterator). Also let us add tests for these result iterators.

File: phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+import org.apache.phoenix.query.QueryServices;
+import org.apache.phoenix.thirdparty.com.google.common.collect.Maps;
+import org.apache.phoenix.util.PhoenixRuntime;
+import org.apache.phoenix.util.ReadOnlyProps;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(NeedsOwnMiniClusterTest.class)
+public class CountRowsScannedIT extends BaseTest {
```
[the remark this answers is not in the document]
Comment (palashc) [reply phoenix-core/src/it/java/org/apache/phoenix/monitoring/CountRowsScannedIT.java:44]: @tkhurana added a test which scans the index

File: phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java (hunk 10 above)
Comment (palashc) [reply phoenix-core-server/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java]: As discussed offline, the changes in this class help with linking the RegionScanner from NonAggregateRegionScannerFactory to the RegionScannerResultIterator, thus bypassing the ResultIterators in between since the current codebase structure makes it difficult to pass the ScannerContext through them on the server side. 
I have also renamed the inner class used here to more meaningful `OrderedResultIteratorWithScannerContext`

**Review by kadirozde — APPROVED** [review]
+1, Thanks!
